File: FurthestInFuturePaging.c
#include <stdio.h>
#include <stdlib.h>

// Implement the furthest-in-future paging algorithm and returns the number of page faults for each instance
int FurthestInFuture(int cacheSize, int numRequests, int* requests) {
    int numFaults = 0;
    int* cache = calloc(cacheSize, sizeof(int)); // Allocate memory for the cache array
    int* future = calloc(cacheSize, sizeof(int)); // Allocate memory for the future array
    
    // Loop through each page request
    for (int x = 0; x < numRequests; x++) {
        int page = requests[x];
        
        // Check if page is already in cache
        int inCache = 0;
        for (int j = 0; j < cacheSize; j++) {
            if (cache[j] == page) {
                inCache = 1;
                break;
            }
        }
        
        // If page is not in cache, add it and evict the page with the furthest future access
        if (!inCache) {
            numFaults++; // Increment the number of page faults
            
            // Find the page in cache with the furthest future access
            int maxFuture = -1;
            int evictIndex = -1;
            for (int y = 0; y < cacheSize; y++) {
                future[y] = numRequests; // Initialize to furthest future
                for (int z = x+1; z < numRequests; z++) {
                    if (requests[z] == cache[y]) {
                        future[y] = z;
                        break;
                    }
                }
                if (future[y] > maxFuture) {
                    maxFuture = future[y];
                    evictIndex = y;
                }
            }
            
            // Evict page with furthest future access and add new page to cache
            cache[evictIndex] = page;
        }
    }
    
    // Free the dynamically allocated memory and return the number of page faults
    free(cache);
    free(future);
    return numFaults;
}

int main() {
    int numInstances;
    scanf("%d", &numInstances); // User input for number of instances
    int* numFaults = malloc(numInstances * sizeof(int)); // Allocate memory for numFaults array to store the number of page faults for each instance
    
    // Loop through each instance
    for (int x = 0; x < numInstances; x++) {
        int cacheSize, numRequests;
        scanf("%d %d", &cacheSize, &numRequests); // User input for cache size and number of page requests
        int* requests = malloc(numRequests * sizeof(int)); // Allocate memory for the requests array

        for (int y = 0; y < numRequests; y++) {
            scanf("%d", &requests[y]); // Read each page request from standard input
        }
        
        numFaults[x] = FurthestInFuture(cacheSize, numRequests, requests); // Run the furthest-in-future paging algorithm and store the number of page faults
        
        free(requests); // Free the dynamically allocated memory
    }

    // Print the number of page faults for each instance
    for (int x = 0; x < numInstances; x++) {
        printf("%d\n", numFaults[x]);
    }

    free(numFaults); // Free the dynamically allocated memory
    
    return 0;
}

File: FurthestInFuturePaging.java
import java.util.*;

public class FurthestInFuturePaging {
    // Calculates the number of page faults using the Furthest-In-Future Paging algorithm
    public static int furthestInFuturePaging(int cacheSize, int[] requests) {
        // Initialize a hashset to store pages in the cache
        Set<Integer> cache = new HashSet<>();
        // Initialize a hashmap to store the next occurrence of each page
        Map<Integer, Integer> futureOccurrence = new HashMap<>();
        // Initialize a variable to keep track of the number of page faults
        int numFaults = 0;

        // Iterate through each request
        for (int x = 0; x < requests.length; x++) {
            int page = requests[x];

            // Check if the page is already in the cache
            if (!cache.contains(page)) {
                numFaults++;

                // If the cache is full, remove the page with the furthest future occurrence
                if (cache.size() == cacheSize) {
                    int pageToRemove = -1;
                    int maxFutureOccurrence = -1;

                    for (int cachedPage : cache) {
                        int futureIndex = futureOccurrence.get(cachedPage);
                        if (futureIndex > maxFutureOccurrence) {
                            maxFutureOccurrence = futureIndex;
                            pageToRemove = cachedPage;
                        }
                    }

                    cache.remove(pageToRemove);
                    futureOccurrence.remove(pageToRemove);
                }

                // Add the new page to the cache
                cache.add(page);
            }

            // Update the future occurrence of the current page
            futureOccurrence.put(page, findNextOccurrence(requests, x));
        }

        // Return the number of page faults
        return numFaults;
    }

    // Finds the next occurrence of a page
    public static int findNextOccurrence(int[] requests, int currentIndex) {
        int nextPage = requests[currentIndex];

        for (int x = currentIndex + 1; x < requests.length; x++) {
            if (requests[x] == nextPage) {
                return x;
            }
        }

        return Integer.MAX_VALUE;
    }

    public static void main(String[] args) {
        // Initialize a scanner to read input from the console
        Scanner sc = new Scanner(System.in);
        // Read user input for the number of test cases
        int numTests = sc.nextInt();
        // Initialize an array to store the number of page faults for each test case
        int numFaults[] = new int[numTests];

        // Iterate through each test case
        for (int x = 0; x < numTests; x++) {
            // Read user input for the cache size and the number of requests for the current test case
            int cacheSize = sc.nextInt();
            int numRequests = sc.nextInt();
            // Initialize an array to store the requests for the current test case
            int[] requests = new int[numRequests];

            // Read user input for the current test case
            for (int y = 0; y < numRequests; y++) {
                requests[y] = sc.nextInt();
            }

            // Calculate the number of page faults for the current test case
            numFaults[x] = furthestInFuturePaging(cacheSize, requests);
        }

        // Close the scanner
        sc.close();

        // Print the number of page faults for each test case
        for(int x = 0; x < numTests; x++) {
            System.out.println(numFaults[x]);
        }
    }

}
